# Исследование надежности заемщиков

## Знакомство с данными

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv('/datasets/data.csv')

In [5]:
display(data.head(10))

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

**Проверим количество пропущенных значений**

In [7]:
print(data.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


В двух столбцах есть пропущенные значения. Один из них — days_employed. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — total_income — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type

In [9]:
for d in data['income_type'].unique(): 
    data.loc[(data['income_type']==d) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type']==d), 'total_income'].median()
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

**Обработаем аномальные значения в столбце days_employed, применим метод abs()**

In [11]:
data['days_employed'] = data['days_employed'].abs()
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   Среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

**Выведем медианное значение для каждого типа занятости**

In [12]:
days_employed_median = data.groupby('income_type')['days_employed'].median()
days_employed_median

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть.

**Выведем уникальные значения столбца children**

In [13]:
print(data['children'].unique())

[ 1  0  3  2 -1  4 20  5]


В столбце children есть два аномальных значения. Удалим их

In [15]:
data = data.loc[(data['children']!=-1)&(data['children']!=20)]
print(data['children'].unique())

[1 0 3 2 4 5]


**Заполним пропуски в столбце days_employed медианными значениями по каждому типу занятости income_type**

In [17]:
for d in data['income_type'].unique():
    data.loc[(data['income_type'] == d)&(data['days_employed'].isna()),'days_employed']=\
    data.loc[(data['income_type'] == d), 'days_employed'].median()
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Проверим пропущенные значения**

In [18]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

**Изменим вещественный тип данных в столбце total_income на целочисленный с помощью метода astype().**

In [19]:
data['total_income']=data['total_income'].astype('int')

**Обработаем неявные дубликаты в столбце education**

In [20]:
data['education']=data['education'].str.lower()

In [21]:
display(data.duplicated().sum())
data=data.drop_duplicates()
display(data.head())
display(data.duplicated().sum())

71

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

0

**Создадим total_income_category с категориями:**
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'

Создадим функцию:

In [22]:
def categorize_income(total_income):
    if total_income<=30000:
        return 'E'
    if total_income<=50000:
        return 'D'
    if total_income<=200000:
        return 'C'
    if total_income<=1000000:
        return 'B'
    return 'A'

In [23]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем перечень уникальных значений методом unique()

In [28]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:
* 'операции с автомобилем',
* 'операции с недвижимостью',
* 'проведение свадьбы',
* 'получение образования'.

In [29]:
def categorize_purpose(index):
    if index == 'приобретение автомобиля' or index =='на покупку подержанного автомобиля' or index =='на покупку своего автомобиля' or index =='автомобили' or index =='сделка с подержанным автомобилем' or index =='автомобиль' or index =='свой автомобиль' or index =='сделка с автомобилем' or index =='на покупку автомобиля':
        return 'операции с автомобилем'
    if index == 'покупка жилья' or index =='операции с жильем' or index =='покупка жилья для семьи' or index =='покупка недвижимости' or index =='покупка коммерческой недвижимости' or index =='покупка жилой недвижимости' or index =='строительство собственной недвижимости' or index =='недвижимость' or index =='строительство недвижимости' or index =='операции с коммерческой недвижимостью' or index =='строительство жилой недвижимости' or index =='жилье' or index =='операции со своей недвижимостью' or index =='покупка своего жилья' or index =='операции с недвижимостью' or index =='покупка жилья для сдачи' or index =='ремонт жилью':
        return 'операции с недвижимостью'
    if index == 'сыграть свадьбу' or index =='на проведение свадьбы' or index =='свадьба':
        return 'проведение свадьбы'
    return 'получение образования'

In [30]:
data['purpose_category']=data['purpose'].apply(categorize_purpose)
display(data)

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

## Исследовательский анализ

**Изучим есть ли зависимость между количеством детей и возвратом кредита в срок**

In [31]:
stat=data.pivot_table(index=['children'], values='debt', aggfunc={'sum','count','mean'})
stat['mean']=stat['mean']*100
stat.reset_index()

children    count      mean     sum
0         0  14091.0  7.543822  1063.0
1         1   4808.0  9.234609   444.0
2         2   2052.0  9.454191   194.0
3         3    330.0  8.181818    27.0
4         4     41.0  9.756098     4.0
5         5      9.0  0.000000     0.0

**Вывод:** Клиенты, не имеющие детей, лучше покрывают свои крелитные обязательства 7,54%(только 1063 из 14091 имели задолженности по кредитам) в сравнении с семьями с 1-2 детьми это 9,23% и 9,45%.

**Выясним есть ли зависимость между семейным положением и возвратом кредита в срок**

In [32]:
stat_by_status = data.pivot_table(index=['family_status'], values = 'debt', aggfunc={'count','sum','mean'})
stat_by_status['mean'] = stat_by_status['mean']*100
stat_by_status.reset_index()

family_status    count      mean    sum
0  Не женат / не замужем   2796.0  9.763948  273.0
1              в разводе   1189.0  7.064760   84.0
2         вдовец / вдова    951.0  6.624606   63.0
3       гражданский брак   4134.0  9.313014  385.0
4        женат / замужем  12261.0  7.560558  927.0

**Вывод:** Лучше погашают свои кредитные обязательства клиенты из категории в женат/замужем - 7,56%(при количестве клиентов 12261 только у 927 были задолженности по кредитам). Хуже всего отдают кредиты клиенты в категориях граждаснкий брак (9,31%) и не женат/не замужем (9,76%)

**Выясним есть ли зависимость между уровнем дохода и возвратом кредита в срок**

In [33]:
data_by_income = data.pivot_table(index=['total_income_category'], values='debt', aggfunc={'sum','count','mean'})
data_by_income['mean']=data_by_income['mean']*100
data_by_income.reset_index()

total_income_category    count      mean     sum
0                     A     25.0  8.000000     2.0
1                     B   5014.0  7.060231   354.0
2                     C  15921.0  8.498210  1353.0
3                     D    349.0  6.017192    21.0
4                     E     22.0  9.090909     2.0

**Вывод:** Кредитополучатели, зарабатывающие в диапазоне "В" лучше погашают свои кредитные обязательства(7,06%) против кредитополучателей в категории "С" - 8,49%. Разница между этими двумя категориями почти 1,5%

**Проверим как разные цели кредита влияют на его возврат в срок**

In [34]:
data_by_purpose = data.pivot_table(index=['purpose_category'], values = 'debt', aggfunc={'sum', 'count', 'mean'})
data_by_purpose['mean']=data_by_purpose['mean']*100
data_by_purpose.reset_index()

purpose_category    count      mean    sum
0    операции с автомобилем   4279.0  9.347978  400.0
1  операции с недвижимостью  10751.0  7.255139  780.0
2     получение образования   3988.0  9.252758  369.0
3        проведение свадьбы   2313.0  7.911803  183.0

**Вывод:** Лучше всего ситуация по задолженностям в категории "операции с недвижимостью" - 7,25% (при рассмотрении 10751 клиентов только у 780 были задолженности). Хуже всего ситуация обстоит по категориям "операции с автомобилем" и "получение образования" - это 9,34% и 9,25% соответственно. На мой взгляд самые надежные кредиторы те, кто берет ипотеку связан с постепенным удорожанием такого актива как недвижимость, ежегодная инфляция и др. причины, в то время как такой актив как автомобиль ежегодно теряет в своей стоимости + ежемесячные расходы на автомобиль(топливо, текущий ремонт, страховка и пр), а также возможные непредвиденные обстоятельства(аварии и пр)

Вывод:

При анализе данной выгрузки было проанализировано 21525 клиентов по различным показателям. 

Выполнена предобработка данных. В датасете имелись пропуски и дубликаты. Причины пропусков в данных могут являться как технические ошибки при выгрузке данных, так и несвоевременное предоставленные данных клиентами. Использование в пропусках медианных значений гораздо лучше отражает реальную картину на рынке труда. Заполнение пропуском средними значениями по таким типам занятости как "сотрудник" и "госслужащий" сильно бы завысило показатели по доходу клиентов. В данном вопросе считаю, что лучше брать заниженные показатели для расчета получения кредита клиентом банка.

У клиентов с 1-2 детьми процент задолженности по кредитным обязательствам выше (от 9,23% до 9,45%) чем у клиентов без детей(7,54%). При анализе 12261 клиента в категории женат/замужем выяснилось, что данная категория лучше погашает свои кредитные обязательства в сравнении с категорией не женат/не замужем - 9,76%. Также категрия клиентов, зарабатывающих в диапазоне от 200 тыс до 1 млн ответственнее относятся к своим кредитным обязательствам. Доля клиентов в этой категории, имеющих задолженность, всего 7,06% по сравнению с категорией клиентов, зарабатывающих в диапазоне от 50 тыс до 200 тыс. Таких клиентов примерно на 1,5% больше, а именно 8,49%. Категория клиентов, приобретащих недвижимость, справляется с погашением своих задолженностей лучше из 10751 только у 780 были задолженности по возвратам кредитов это 7,25%. По категории "операции с автомобилем" и "получение образования" процент гораздо выше от 9,34% до 9,25% (в этих категориях рассматрирвалось чуть более 8 тыс клиентов суммарно). Для более точного анализа выгрузки на мой взгляд лучше исключить данные, имеющие пропуски в таких важных показателях как ежемесячный доход и общий трудовой стаж в днях. Ведь от этих данных зависит понимание благонадежности клиента, так и сумма/расчет получаемого кредита. 